include/ignition/rendering/CameraLens.hh

```diff
@@ -0,0 +1,190 @@
+      /// \return True if this->Type() == MFT_CUSTOM
+      public: bool IsCustom() const;
+
+      /// \brief Gets c1 parameter
+      /// \return c1 parameter
+      public: double C1() const;
+
+      /// \brief Gets c2 parameter
+      /// \return c2 parameter
+      public: double C2() const;
+
+      /// \brief Gets c3 parameter
+      /// \return c3 parameter
+      public: double C3() const;
+
+      /// \brief Gets f parameter
+      /// \return f parameter
+      public: double F() const;
+
+      /// \brief Gets angle transform function
+      /// \return Angle transform function
+      public: AngleFunctionType AngleFunction() const;
+
+      /// \brief Gets cut off angle
+      /// \return Cut off angle
+      public: double CutOffAngle() const;
+
+      /// \brief Checks if image should be scaled to fit horizontal FOV
+      /// \return True if the image will be scaled
+      public: bool ScaleToHFOV() const;
+
+      /// \brief Set lens projection type
+      /// \param[in] _type Lens projection / mapping function type
+      public: void SetType(MappingFunctionType _type);
+
+      /// \brief Sets c1 parameter
+      /// \param[in] _c c1 parameter
+      public: void SetC1(double _c);
+
+      /// \brief Sets c2 parameter
+      /// \param[in] _c c2 parameter
+      public: void SetC2(double _c);
+
+      /// \brief Sets c3 parameter
+      /// \param[in] _c c3 parameter
+      public: void SetC3(double _c);
+
+      /// \brief Sets f parameter
+      /// \param[in] _f f parameter
+      public: void SetF(double _f);
+
+      /// \brief Sets angle transform function
+      /// \param[in] _fun Angle transform function
+      public: void SetAngleFunction(AngleFunctionType _fun);
+
+      /// \brief Sets cut-off angle
+      /// \param[in] _angle cut-off angle
+      public: void SetCutOffAngle(double _angle);
+
+      /// \brief Sets whether the image should be scaled to fit horizontal FOV
+      /// If True, the projection will compute a new focal length for achieving
+      /// the desired FOV
+      /// \param[in] _scale true if it should,
+      ///   note: c1 and f parameters are ignored in this case
+      public: void SetScaleToHFOV(bool _scale);
+
+      /// \brief Apply mapping function to input number
+      /// \return Result of the application
+      public: float ApplyMappingFunction(float _f) const;
+
+      /// \internal
+      /// \brief Get mapping function as vector3d
+      /// return unit vector, either unit x, y, or z.
+      public: math::Vector3d MappingFunctionAsVector3d() const;
+
+      /// \brief Assignment operator
+      /// \param[in] _other The other camera lens
+      public: CameraLens &operator=(const CameraLens &_other);
+
+      /// \internal
+      /// \brief Converts projection type from one of presets to `custom`
+      private: void ConvertToCustom();
+
+      /// \internal
+      /// \brief Private data pointer
+      private: std::unique_ptr<CameraLensPrivate> dataPtr;
```

Comment:
I think we could add `ignition-utils` as a dependency so we can use the `ImplPtr`? I'll open an issue to do this in parallel, without blocking

Reply:
this dependency has been added in #492, so I recommend using `ImplPtr` in this new class

Reply:
switched to use `ImplPtr` in 21d9f79